src/Week7/HW7.hs

```diff
@@ -0,0 +1,240 @@
+import Control.Comonad (extract)
+
+-- Exercise 1 -----------------------------------------
+-- Define a function that lifts a regular function into a monad and applies it to the argument of that monad
+liftM' :: Monad m => (a -> b) -> m a -> m b
+liftM' f ma = f <$> ma
+
+-- Alternatively
+liftM'' f ma = ma >>= \a -> return $ f a
+
+-- or even 
+liftM''' f ma = ma >>= (f >=> return)
+-- HLS suggested this >=> operator. Pretty cool :D
+
+-- Define a function that takes in two indices and swaps the elements at those indices in some vector
+swapV :: Int -> Int -> Vector a -> Maybe (Vector a)
+swapV i j v = do
+  let tmp = v !? i
+  let change1 = liftM' ((,)i) (v!?j)
+  let change2 = liftM' ((,)j) tmp
+  let changes = sequence [change1, change2]
+  (v V.//) <$> changes
+-- This could be shortened but I chose the version above because it is (I think) readable  
+
+-- Exercise 2 -----------------------------------------
+-- Implement a function that maps a monadic function across a list
+mapM' :: Monad m => (a -> m b) -> [a] -> m [b]
+mapM' f lst = sequence $ f <$> lst
+
+-- Use mapM to define a function that takes in a list of indices and a vector and returns a list of the elements at those indices in the Maybe monad
+getElts :: [Int] -> Vector a -> Maybe [a]
+getElts lst v = mapM' (v V.!?) lst
+
+-- Exercise 3 -----------------------------------------
+
+type Rnd a = Rand StdGen a
+
+-- Use the randomness monad to produce a random element of a vector
+randomElt :: Vector a -> Rnd (Maybe a)
+randomElt v = case length v of
+  0 -> return Nothing
+  l -> (v V.!?) <$> getRandomR (0, l-1)
+
+-- Exercise 4 -----------------------------------------
+-- Write a function that outputs a vector of specified length with random elements
+randomVec :: Random a => Int -> Rnd (Vector a)
+randomVec l = sequence $ V.generate l (const getRandom)
+
+-- Same as above, but with elements within a range
+randomVecR :: Random a => Int -> (a, a) -> Rnd (Vector a)
+randomVecR l (a,b) = sequence $ V.generate l (const $ getRandomR (a,b))
+
+-- Exercise 5 -----------------------------------------
+--  Implement the Fisher-Yates algorithm to shuffle the elements of a vector
+
+swapV' :: Int -> Int -> Vector a -> Vector a
+swapV' a b v = Ma.fromJust $ swapV a b v
+--swapV' = Ma.fromJust . swapV 
+-- why can we not leave out the parameters, as in the line above?
+
+shuffle :: Vector a -> Rnd (Vector a)
+shuffle vt = helper (pure vt) (length vt - 1)
+  where helper v 0 = v
+        helper v n = helper (swapV' n <$> getRandomR (0, n-1) <*> v) (n - 1)
+
+-- Exercise 6 -----------------------------------------
+-- Implement a function that partitions a vector around the element at a given index
+
+-- Appending an element to a vector is O(N), so we first convert the input to a list (to use O(1) cons) and then convert the result back to a vector at the end
+partitionAt :: Ord a => Vector a -> Int -> (Vector a, a, Vector a)
+partitionAt v i = toTupleV $ helper [] (v!i) [] i (V.toList v)
```

Comment:
Using https://hackage.haskell.org/package/vector-0.12.3.0/docs/Data-Vector.html#v:splitAt would be preferrable.
Also this function is partial (will break with empty vector and for out-of-bounds index).